### ISSUE TITLE
Issue with the Form Digest in SPFx

### ISSUE THREAD

**Issue by martinlingstuyl** [pnp/pnpjs#2379]
### Category
- [ ] Enhancement
- [x] Bug
- [ ] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: 3.5.1

Please specify what version(s) of SharePoint you are targeting: SharePoint Online

### Expected / Desired Behavior / Question
If I'm calling SharePoint using PnP Js from SPFx I want a valid FormDigest to be included if necessary, to avoid 403 errors

### Observed Behavior
I've seen a number of closed issues regarding the expiring FormDigest. Despite that I still ran into 403 errors caused by an invalid FormDigest.

Strangely these issues seemed to occur after page refreshes. What I did was as follows:
1) Leave a SharePoint page with webpart open for more than an hour.
2) Interact with the webpart. (works fine)
3) Refresh the page
4) Iinteract with the same webpart => 403

It appears the Form Digest Value in the `pageContext.legacyPageContext` object is not always valid and that this is not checked when initializing SPFI.

I worked around this by adding the `RequestDigest()` behavior to the SPFI tunnel explicitly. This triggers a `contextinfo` api call when loading the SPFx webpart for the first time, thereby fixing my issue.

Check out the code in my SPFx service:

```typescript
private _sp: SPFI;

public constructor(serviceScope: ServiceScope) {
    serviceScope.whenFinished(() => {
        const pageContext = serviceScope.consume(PageContext.serviceKey);
        this._sp = spfi().using(RequestDigest(), SPFx({ pageContext }));
    });
}
``` 

I don't think this is what you would expect though. `spfi().using(SPFx({ pageContext }))` should be enough, right? 
So maybe this is a pointer to something that can be fixed.

**Comment by juliemturner** [comment]
If you'll see this issue #2304 we did address that issue a while ago, there was then an issue with Safari which we reversed as the issue states but yes we do check the legacyPageContext... specifically this is what we do https://github.com/pnp/pnpjs/blob/af3d8131224de6d8c8925576f091ba20608098d8/packages/sp/behaviors/spfx.ts#L34
Also, here's the docs on the same... https://pnp.github.io/pnpjs/sp/behaviors/#spfx

**Comment by martinlingstuyl** [comment]
I had read that and seen the changes, yes.

I'm busy checking if I can set up a basic example to show what's going on. 👍

**Comment by juliemturner** [comment]
I tested it just now... I have a sample solution that updates file names, and it worked fine after waiting about 1:15 mins... so, I would need a test project as you mentioned to figure out what the issue is.

**Comment by martinlingstuyl** [comment]
I will probably share something tomorrow 👍

**Comment by martinlingstuyl** [comment]
Hi @juliemturner, OK, so I've created a sample solution, you can find it here:

https://github.com/martinlingstuyl/pnpjs-requestdigest

The behavior only occurs in the SPFx extension. The webpart seems to work fine.

I've added a gif showing what's happening on my end + reproduction steps. If you have any questions, let me know.

**Comment by juliemturner** [comment]
Ok, I finally got around to debugging this and it turns out you've managed to find a but in SPFx. This is not an issue with PnPjs.  The issue is that for some reason when you're constructor runs and asks the ServiceScope to consume the page context the page context you're getting back is the one from the previous page refresh... ergo the token in the case where prior to the refresh the token was expired ends up getting through. The reason your work around works is because you're bypassing the code that uses the page token from the page. What I'm not sure without doing more digging is if this "ServiceScope" bug is a bug for all different SPFx project types or just command sets... I suspect all...

So long story short what I would suggest you do is strip this repo you shared with me down to just the service and then reproduce what I show in the screen shot before (maybe wait a bit longer between the refreshes) and submit it to sp-dev-docs for support and you can reference this issue.

![image](https://user-images.githubusercontent.com/7570936/187780709-2c2b9d7e-c387-4b40-815c-3df14dff5d04.png)

**Comment by martinlingstuyl** [comment]
Thanks @juliemturner for checking this out! I suppose I could have reached that conclusion myself as well. 🙄 I'll certainly go and notify the SPFx team about this.

What do we do with this issue, close it?

**Comment by juliemturner** [comment]
Yep, we can close it.
